File: app/src/main/java/com/babic/filip/movieshack/ui/home/HomeActivity.java
package com.babic.filip.movieshack.ui.home;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import com.babic.filip.movieshack.R;
import com.babic.filip.movieshack.listener.RefreshablePage;
import com.babic.filip.movieshack.ui.popular.PopularMoviesFragment;
import com.babic.filip.movieshack.ui.topRated.TopRatedMoviesFragment;
import com.babic.filip.movieshack.ui.upcoming.UpcomingMoviesFragment;

public class HomeActivity extends AppCompatActivity implements BottomNavigationView.OnNavigationItemReselectedListener, BottomNavigationView.OnNavigationItemSelectedListener {

    private RefreshablePage refreshablePage = null;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);

        BottomNavigationView bottomNavigation = findViewById(R.id.bottomNavigation);

        bottomNavigation.setOnNavigationItemSelectedListener(this);
        bottomNavigation.setOnNavigationItemReselectedListener(this);
    }

    private void showUpcoming() {
        UpcomingMoviesFragment fragment = new UpcomingMoviesFragment();
        refreshablePage = fragment;

        replacePage(fragment);
    }

    private void showTopRated() {
        TopRatedMoviesFragment fragment = new TopRatedMoviesFragment();
        refreshablePage = fragment;

        replacePage(fragment);
    }

    private void showPopular() {
        PopularMoviesFragment fragment = new PopularMoviesFragment();
        refreshablePage = fragment;

        replacePage(fragment);
    }

    private void replacePage(Fragment fragment) {
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.frameContainer, fragment)
                .commit();
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.upcomingMovies: {
                showUpcoming();
                break;
            }

            case R.id.topMovies: {
                showTopRated();
                break;
            }

            case R.id.popularMovies: {
                showPopular();
                break;
            }

            default:
                break;
        }

        return true;
    }

    @Override
    public void onNavigationItemReselected(@NonNull MenuItem item) {
        if (refreshablePage != null) {
            refreshablePage.refresh();
        }
    }
}

File: app/src/main/java/com/babic/filip/movieshack/ui/home/HomeActivity.kt
package com.babic.filip.movieshack.ui.home

import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem

import com.babic.filip.movieshack.R
import com.babic.filip.movieshack.listener.RefreshablePage
import com.babic.filip.movieshack.ui.popular.PopularMoviesFragment
import com.babic.filip.movieshack.ui.topRated.TopRatedMoviesFragment
import com.babic.filip.movieshack.ui.upcoming.UpcomingMoviesFragment
import kotlinx.android.synthetic.main.activity_home.*

class HomeActivity : AppCompatActivity(), BottomNavigationView.OnNavigationItemReselectedListener, BottomNavigationView.OnNavigationItemSelectedListener {

    private var refreshablePage: RefreshablePage? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)

        bottomNavigation.setOnNavigationItemSelectedListener(this)
        bottomNavigation.setOnNavigationItemReselectedListener(this)
        bottomNavigation.selectedItemId = R.id.upcomingMovies

        showUpcoming()
    }

    private fun showUpcoming() {
        val fragment = UpcomingMoviesFragment()
        refreshablePage = fragment

        replacePage(fragment)
    }

    private fun showTopRated() {
        val fragment = TopRatedMoviesFragment()
        refreshablePage = fragment

        replacePage(fragment)
    }

    private fun showPopular() {
        val fragment = PopularMoviesFragment()
        refreshablePage = fragment

        replacePage(fragment)
    }

    private fun replacePage(fragment: Fragment) = supportFragmentManager
            .beginTransaction()
            .replace(R.id.frameContainer, fragment)
            .commit()

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.upcomingMovies -> showUpcoming()
            R.id.topMovies -> showTopRated()
            R.id.popularMovies -> showPopular()
            else -> Unit
        }

        return true
    }

    override fun onNavigationItemReselected(item: MenuItem) = refreshablePage?.refresh() ?: Unit
}
